Enhance the help command to print info about a given unit

eminence opened the pull request:
This PR starts to implement the enhanced help suggestion in #238.

At the moment, it looks like this:

```
>>> help megabyte
  Byte
  https://en.wikipedia.org/wiki/Byte
  Aliases: B, byte, bytes, Byte, Bytes, octet, octets, Octet, Octets
  A unit of [DigitalInformation]
  
  1 megabyte = 1_000_000 byte
  1 byte = 8 bit
```

```
>>> help m
  Metre
  https://en.wikipedia.org/wiki/Metre
  Aliases: metre, metres, meter, meters, m
  A unit of [Length]
  
  metre is a base unit
```


CC #238

eminence commented on numbat/src/lib.rs:
```diff
@@ -190,6 +193,74 @@ impl Context {
+        let reg = self.interpreter.get_unit_registry();
+
+        if let PrefixParserResult::UnitIdentifier(_span, prefix, _, full_name) =
+            self.prefix_transformer.prefix_parser.parse(keyword)
```
I'm not sure if using the `prefix_parser` is the best way to do this (since it required me making it `pub`)

sharkdp replied:
I think it's fine

eminence commented on numbat/src/unit.rs:
```diff
@@ -42,6 +42,18 @@ impl UnitIdentifier {
         matches!(self.kind, UnitKind::Base)
     }
 
+    pub fn unit_and_factor(&self) -> BaseUnitAndFactor {
```
The idea behind this function is an analog to `base_unit_and_factor` except that it doesn't convert it into the base unit.

eminence commented:
The second commit adds support for attaching metadata to local variables and into the `help` command:

```
>>> help planck_constant
  Variable: planck_constant
  Planck constant
  https://en.wikipedia.org/wiki/Planck_constant
  Aliases: planck_constant, ℎ
  
      = 6.62607e-34 J/Hz
  ```

sharkdp commented on numbat-cli/src/main.rs:
```diff
@@ -370,6 +370,15 @@ impl Cli {
                                 println!();
                             }
                             _ => {
+                                if let Some(keyword) = line.strip_prefix("help ") {
```
Maybe
```suggestion
                                if let Some(keyword) = line.strip_prefix("help").trim() {
```

sharkdp commented on numbat/modules/physics/constants.nbt:
```diff
@@ -4,77 +4,95 @@ use units::si
 let speed_of_light: Velocity = 299_792_458 m / s
-let c = speed_of_light
 
 # The Newtonian constant of gravitation
+@url("https://en.wikipedia.org/wiki/Gravitational_constant")
```
```suggestion
@name("Newtonian constant of gravitation")
@url("https://en.wikipedia.org/wiki/Gravitational_constant")
```

sharkdp commented on numbat/modules/physics/constants.nbt:
```diff
@@ -4,77 +4,95 @@ use units::si
 ### Physics constants
 
 # The speed of light in vacuum
```
```suggestion
```

sharkdp commented on numbat/src/bytecode_interpreter.rs:
```diff
@@ -185,13 +193,29 @@ impl BytecodeInterpreter {
-                    depth: 0,
-                });
+
+                // For variables, we ignore the prefix info and only use the names
```
Should we allow disallow adding prefix information in the parser?

eminence replied:
Agreed.  I just pushed a commit that disallows this in the parser

sharkdp reviewed (commented):
Really nice — thank you very much!

I added some additional metadata for math constants.

eminence commented:
Ok, given that I didn't see any objections to the approach taken here, I think this is ready for a fuller review.  I want to take another pass at the output formatting of the help output (I welcome any and all comments on this)

sharkdp commented on numbat-cli/src/main.rs:
```diff
@@ -370,6 +370,15 @@ impl Cli {
                                 println!();
                             }
                             _ => {
+                                if let Some(keyword) = line.strip_prefix("help") {
```
I was thinking about using `help <identifier>` again, and wondered if we should maybe introduce a new `info <identifier>` instead? I think it would reflect a bit better what the command does. And it would keep `help <topic>` free if we ever want to enhance the in-program help text for things like `help commands`, `help syntax` or similar.

What do you think?

eminence replied:
Sure, I like that suggestion.  I've just pushed a commit that renames the command to `info`

sharkdp commented on numbat/src/bytecode_interpreter.rs:
```diff
@@ -359,6 +384,20 @@ impl BytecodeInterpreter {
         self.locals.len() - 1
     }
+
+    pub fn get_defining_unit(&self, unit_name: &str) -> Option<&Unit> {
+        self.unit_name_to_constant_index
+            .get(unit_name)
+            .and_then(|idx| self.vm.constants.get(*idx as usize))
+            .and_then(|constant| match constant {
+                Constant::Unit(u) => Some(u),
+                _ => None,
+            })
+    }
```
Nice!

sharkdp commented on numbat/src/decorator.rs:
```diff
@@ -66,3 +66,15 @@ pub fn url(decorators: &[Decorator]) -> Option<String> {
     None
 }
+
+pub fn contains_aliases_with_prefixes(decorates: &[Decorator]) -> bool {
+    for decorator in decorates {
```
minor:
```suggestion
pub fn contains_aliases_with_prefixes(decorators: &[Decorator]) -> bool {
    for decorator in decorators {
```

or is "decorates" the correct term here? :-)

sharkdp commented on numbat/src/lib.rs:
```diff
@@ -253,6 +256,106 @@ impl Context {
+                        + m::nl();
+                }
+
+                help += m::text("A unit of [") + md.readable_type + m::text("]") + m::nl();
```
Maybe we could special-case `Scalar`/dimensionless quantities here. For example: `info million` prints:
```
A unit of [Scalar]
```
Would be great to print sth. like "A dimensionless unit".

eminence replied:
do you see any value in keeping the word "Scalar" just to make sure users are familiar with it?   For example, the output of `type(5)` is `Scalar` and `Scalar` is used when definition functions that only scalar values.  Maybe we could do:

```
A dimensionless unit ([Scalar])
```

sharkdp replied:
Yeah, sounds good

sharkdp commented on numbat/src/lib.rs:
```diff
@@ -253,6 +256,106 @@ impl Context {
+            return help;
+        }
+
+        m::text("Not found")
```
It would be fantastic if we could also use this for types, i.e. do `info ElectricCharge` and get information about how it is defined and what the base representation is. But I guess we can leave this for another PR.

sharkdp commented on numbat/src/lib.rs:
```diff
@@ -253,6 +256,106 @@ impl Context {
+                let mut help =
+                    m::text("Unit: ") + m::unit(md.name.as_deref().unwrap_or(keyword)) + m::nl();
+                if let Some(url) = &md.url {
+                    help += m::string(url) + m::nl();
```
I think we could maybe put the URL after the name? And similar for variables.
```
>>> info C
  Unit: Coulomb (https://en.wikipedia.org/wiki/Coulomb)
  Aliases: coulomb, coulombs, C
  A unit of [ElectricCharge]
  
  1 coulomb = 1 A x s
```

sharkdp commented on numbat/src/lib.rs:
```diff
@@ -253,6 +256,106 @@ impl Context {
+                    if !prefix.is_none() {
+                        help += m::nl()
+                            + m::value("1 ")
+                            + m::type_identifier(keyword)
```
This should be
```suggestion
                            + m::unit(keyword)
```

sharkdp commented on numbat/src/lib.rs:
```diff
@@ -253,6 +256,106 @@ impl Context {
+                            + m::text(" = ")
+                            + m::value(prefix.factor().pretty_print())
+                            + m::space()
+                            + m::type_identifier(&full_name);
```
Same here:
```suggestion
                            + m::unit(&full_name);
```

sharkdp commented on numbat/src/lib.rs:
```diff
@@ -253,6 +256,106 @@ impl Context {
+                    if let Some(BaseUnitAndFactor(prod, num)) = x {
+                        help += m::nl()
+                            + m::value("1 ")
+                            + m::type_identifier(&full_name)
```
```suggestion
                            + m::unit(&full_name)
```

sharkdp commented on numbat/src/lib.rs:
```diff
@@ -253,6 +256,106 @@ impl Context {
+                            + m::text(" = ")
+                            + m::value(num.pretty_print())
+                            + m::space()
+                            + prod.pretty_print_with(|f| f.exponent, 'x', '/', true);
```
I guess this always uses `type_identifier` for the markup? In that case, maybe we can pass the markup style as an argument (see `FormatType`)? Or pass m::type_identifier/m:unit as a function argument?

sharkdp commented on numbat/src/lib.rs:
```diff
@@ -253,6 +256,106 @@ impl Context {
+                            + prod.pretty_print_with(|f| f.exponent, 'x', '/', true);
+                    } else {
+                        help +=
+                            m::nl() + m::type_identifier(&full_name) + m::text(" is a base unit");
```
```suggestion
                            m::nl() + m::unit(&full_name) + m::text(" is a base unit");
```

sharkdp reviewed (commented):
Really cool. We should also think about the web version. I can also make the necessary changes if you can't run it, for example.


Maybe we can also add a simple integration test in `numbat-cli/tests/integration.rs` (see similar test for `help`: `help_text`).

eminence commented:
Ok I think I addressed all the review comments.  Here's a screenshot showing how it looks now:

![image](https://github.com/sharkdp/numbat/assets/402454/caca1370-a488-4f16-bc9f-eb679aa4583f)

(please also make sure 341ca6d69b979968b82fd8071cf57a7422a40fa6 is OK)

sharkdp commented:
> (please also make sure [341ca6d](https://github.com/sharkdp/numbat/commit/341ca6d69b979968b82fd8071cf57a7422a40fa6) is OK)

:+1:

eminence commented:
New commits:  a cleanup on how variables are shown, an integration test, and support in the wasm version
